nav2_smac_planner/src/node_lattice.cpp

```diff
@@ -59,29 +58,102 @@ void LatticeMotionTable::initMotionModel(
   reverse_penalty = search_info.reverse_penalty;
   current_lattice_filepath = search_info.lattice_filepath;
 
-  // TODO(Matt) read in file, precompute based on orientation bins for lookup at runtime
-  // file is `search_info.lattice_filepath`, to be read in from plugin and provided here.
+  latticeMetadata = getLatticeMetadata(current_lattice_filepath);
+
+  std::ifstream latticeFile(current_lattice_filepath);
+
+  if( !latticeFile.is_open() )
+  {
+    throw std::runtime_error("Could not open lattice file");
+  }
+
+  nlohmann::json j;
+  latticeFile >> j;
+
+  float prevStartAngle = 0; 
+  std::vector<MotionPrimitive> primitives; 
+  nlohmann::json jsonPrimitives = j["primitives"];
+  for(unsigned int i = 0; i < jsonPrimitives.size(); ++i)
+  {
+    MotionPrimitive newPrimitive; 
+    fromJsonToMotionPrimitive(jsonPrimitives[i], newPrimitive); 
+
+    if(prevStartAngle != newPrimitive.start_angle )
+    {
+      motionPrimitives.emplace_back(primitives);
+      primitives.clear(); 
+      prevStartAngle = newPrimitive.start_angle;
+    }
+    primitives.emplace_back(newPrimitive); 
+  }
+  motionPrimitives.emplace_back(primitives); 
+
+  num_angle_quantization = latticeMetadata.number_of_headings;
+  num_angle_quantization_float = static_cast<float>(num_angle_quantization);
+  min_turning_radius = latticeMetadata.min_turning_radius;
+
+  //TODO: Note sure what type of state space this should be 
+  state_space = std::make_unique<ompl::base::DubinsStateSpace>( min_turning_radius );
 
-  // TODO(Matt) create a state_space with the max turning rad primitive within the file
-  // (or another -- mid?)
-  // to use for analytic expansions and heuristic generation. Potentially make both an
-  // extreme and a passive one?
+  for(unsigned int i = 0; i < motionPrimitives.size(); ++i)
+  {
+    for(unsigned int j = 0; j < motionPrimitives[0].size(); ++j)
+    {
+      //TODO: angles in json file need to be converted to radians
+      primitive_headings.emplace_back( motionPrimitives[i][j].start_angle );
+    }
+  }
 
-  // TODO(Matt) populate num_angle_quantization, size_x, min_turning_radius, trig_values,
-  // all of the member variables of LatticeMotionTable
+  //TODO: angles in json file need to be converted to radians
+  for(unsigned int i = 0; i < primitive_headings.size(); ++i)
+  {
+    trig_values.emplace_back( 
+      cos(primitive_headings[i]), 
+      sin(primitive_headings[i]) );
+  } 
 }
 
-MotionPoses LatticeMotionTable::getProjections(const NodeLattice * node)
+MotionPoses LatticeMotionTable::getMotionPrimitives(const NodeLattice * node)
 {
-  return MotionPoses();  // TODO(Matt) lookup at run time the primitives to use at node
+  //Question: Theta is in bin coords, 0-16 so theta does not have to be shifted?
```

Comment:
There are some reasonable questions here:
- Should we be returning only the final point, or should we be returning the full primitive (e.g. each of the points making up the primitive translated to the given point)? That might help later for checking for collisions. 
- What are the coordinates of these primitives? Are they in cells or in meters?

Reply:
I guess having each node store the primitive might be expensive. Also, each primitive would have to be shifted which would add computation. If we were to store the index of the motion primitive in the node we could then pass in the primitive and pose into the collision checker. 

Then we could potentially do some down-sampling within the collision check to reduce the number of points that we would need to check and shift. 

However, the longest primitive in the json file only has 14 points. So it might not make that much of a difference. 

The coordinates of the json file are in meters and radians. So I converted within the function. 

The primitives are in meters and radius, so